File: agent/packages/schema/src/processing.ts
import { Type, type Static } from "@sinclair/typebox";

export const ProcessingKindV1 = Type.Union(
  [
    Type.Literal("drying"),
    Type.Literal("grinding"),
    Type.Literal("forging_alchemy"),
    Type.Literal("extraction"),
  ],
  { description: "plan-lingtian-process-v1 §5.3 四类作物二级加工工艺" },
);
export type ProcessingKindV1 = Static<typeof ProcessingKindV1>;

export const ProcessingSessionDataV1 = Type.Object(
  {
    session_id: Type.String(),
    kind: ProcessingKindV1,
    recipe_id: Type.String(),
    progress_ticks: Type.Integer({ minimum: 0 }),
    duration_ticks: Type.Integer({ minimum: 0 }),
    player_id: Type.String(),
  },
  { additionalProperties: false },
);
export type ProcessingSessionDataV1 = Static<typeof ProcessingSessionDataV1>;

export const FreshnessUpdateV1 = Type.Object(
  {
    item_uuid: Type.String(),
    freshness: Type.Number({ minimum: 0, maximum: 1 }),
    profile_name: Type.String(),
  },
  { additionalProperties: false },
);
export type FreshnessUpdateV1 = Static<typeof FreshnessUpdateV1>;

export const ServerDataProcessingSessionV1 = Type.Intersect([
  Type.Object({ type: Type.Literal("processing_session"), v: Type.Literal(1) }),
  ProcessingSessionDataV1,
]);
export type ServerDataProcessingSessionV1 = Static<typeof ServerDataProcessingSessionV1>;

export const ServerDataFreshnessUpdateV1 = Type.Intersect([
  Type.Object({ type: Type.Literal("freshness_update"), v: Type.Literal(1) }),
  FreshnessUpdateV1,
]);
export type ServerDataFreshnessUpdateV1 = Static<typeof ServerDataFreshnessUpdateV1>;

File: server/src/schema/processing.rs
//! plan-lingtian-process-v1 §5.3 — 作物加工 IPC 镜像。

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ProcessingKindV1 {
    Drying,
    Grinding,
    ForgingAlchemy,
    Extraction,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ProcessingSessionDataV1 {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub active: Option<bool>,
    pub session_id: String,
    pub kind: ProcessingKindV1,
    pub recipe_id: String,
    pub progress_ticks: u32,
    pub duration_ticks: u32,
    pub player_id: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct FreshnessUpdateV1 {
    pub item_uuid: String,
    pub freshness: f32,
    pub profile_name: String,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn processing_session_data_roundtrip_uses_snake_case_kind() {
        let payload = ProcessingSessionDataV1 {
            active: Some(true),
            session_id: "processing:1".to_string(),
            kind: ProcessingKindV1::ForgingAlchemy,
            recipe_id: "forge_ci_she_hao".to_string(),
            progress_ticks: 10,
            duration_ticks: 6_000,
            player_id: "offline:Azure".to_string(),
        };
        let json = serde_json::to_value(&payload).unwrap();
        assert_eq!(json["kind"], "forging_alchemy");
        let decoded: ProcessingSessionDataV1 = serde_json::from_value(json).unwrap();
        assert_eq!(decoded, payload);
    }

    #[test]
    fn freshness_update_bounds_are_runtime_validated_by_callers() {
        let payload = FreshnessUpdateV1 {
            item_uuid: "item:7".to_string(),
            freshness: 0.42,
            profile_name: "drying_v1".to_string(),
        };
        let json = serde_json::to_value(&payload).unwrap();
        assert_eq!(json["profile_name"], "drying_v1");
    }
}
